sdk/identity/Azure.Identity/src/AuthorizationCodeCredential.cs

```diff
@@ -0,0 +1,121 @@
+        /// See https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow for more information.</param>
+        public AuthorizationCodeCredential(string tenantId, string clientId, string clientSecret, string authorizationCode)
+            : this(tenantId, clientId, clientSecret, authorizationCode, null)
+        {
+        }
+
+        /// <summary>
+        /// Creates an instance of the ClientSecretCredential with the details needed to authenticate against Azure Active Directory with a prefetched authorization code.
+        /// </summary>
+        /// <param name="tenantId">The Azure Active Directory tenant (directory) Id of the service principal.</param>
+        /// <param name="clientId">The client (application) ID of the service principal</param>
+        /// <param name="clientSecret">A client secret that was generated for the App Registration used to authenticate the client.</param>
+        /// <param name="authorizationCode">The authorization code obtained from a call to authorize. The code should be obtained with all required scopes.
+        /// See https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow for more information.</param>
+        /// <param name="options">Options that allow to configure the management of the requests sent to the Azure Active Directory service.</param>
+        public AuthorizationCodeCredential(string tenantId, string clientId, string clientSecret, string authorizationCode, IdentityClientOptions options)
+        {
+            if (tenantId is null) throw new ArgumentNullException(nameof(tenantId));
+            if (clientId is null) throw new ArgumentNullException(nameof(clientId));
+            if (clientSecret is null) throw new ArgumentNullException(nameof(clientSecret));
+
+            _authCode = authorizationCode ?? throw new ArgumentNullException(nameof(authorizationCode));
+
+            options ??= new IdentityClientOptions();
+
+            _pipeline = HttpPipelineBuilder.Build(options);
+
+            _confidentialClient = ConfidentialClientApplicationBuilder.Create(clientId).WithHttpClientFactory(new HttpPipelineClientFactory(_pipeline)).WithTenantId(tenantId).WithClientSecret(clientSecret).Build();
+        }
+
+        /// <summary>
+        /// Obtains a token from the Azure Active Directory service, using the specified authorization code authenticate.
+        /// </summary>
+        /// <param name="request">The details of the authentication request.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        /// <returns>An <see cref="AccessToken"/> which can be used to authenticate service client calls.</returns>
+        public override AccessToken GetToken(TokenRequest request, CancellationToken cancellationToken = default)
+        {
+            return GetTokenAsync(request, cancellationToken).GetAwaiter().GetResult();
```

Comment:
Do we track anywhere all the features we need from dependent APIs to remove sync over async at some point?

Reply:
MSAL only supports async at this point. We have brought this up as a general ask to support Sync, but I it's fairly low in priority relative to other asks we have for them. Also, they currently have a hard dependency on HttpClient, which we patch to use our pipeline, so this would have to change in some way.